### PR TITLE
Make codegen I/O-free and agnostic to output location

### PR DESCRIPTION
**Pull request by teh-cmc**
**Commit by commit**

This is necessary refactoring work for the upcoming `attr.rust.custom_crate` attribute, itself necessary for the upcoming serde-codegen support, itself necessary for the upcoming blueprint experimentations as well as #3741.

### Changes

1. The `CodeGenerator` trait as well as all post-processing helpers (gitattributes, orphan detection...) are now I/O-free.
  ```rust
  pub type GeneratedFiles = std::collections::BTreeMap<camino::Utf8PathBuf, String>;
  
  pub trait CodeGenerator {
      fn generate(
          &mut self,
          reporter: &crate::Reporter,
          objects: &crate::Objects,
          arrow_registry: &crate::ArrowRegistry,
      ) -> GeneratedFiles;
  }
  ```
2. All post-processing helpers are now agnostic to the location output.
This is very important as it makes it possible to generate e.g. rust code out of the `re_types` crate without everything crumbling down.
A side-effect is that gitattributes files are now finer-grained.
3. The Rust codegen pass is now crate agnostic: it is driven by the workspace path rather than a specific crate path.
Necessary for the upcoming `attr.rust.custom_crate`.
4. All codegen passes now follow the exact same 4-step structure:
  ```
  // 1. Generate in-memory code files.
  let mut gen = MyGenerator::new();
  let mut files = gen.generate(reporter, objects, arrow_registry);
  // 2. Generate in-memory attribute files.
  generate_gitattributes_for_generated_files(&mut files);
  // 3. Write all in-memory files to disk.
  write_files(&gen.pkg_path, &gen.testing_pkg_path, &files);
  // 4. Remove orphaned files.
  crate::codegen::common::remove_orphaned_files(reporter, &files);
  ```
5. The documentation codegen pass now removes its orphans, which is why some `md` files were removed in this PR.

---

- Unblocks #3741 
- Unblocks https://github.com/rerun-io/rerun/issues/3495

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3888) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3888)
- [Docs preview](https://rerun.io/preview/136f85154039c5e43399ff562b5541c84be94be6/docs) 
- [Examples preview](https://rerun.io/preview/136f85154039c5e43399ff562b5541c84be94be6/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: rerun_cpp/src/rerun/archetypes/.gitattributes
Comment (teh-cmc) [inline rerun_cpp/src/rerun/archetypes/.gitattributes:1]: what the, why is this one all crazy oO

File: crates/re_types_builder/src/codegen/cpp/mod.rs [hunk 1]
```diff
@@ -151,24 +141,22 @@ impl CppCodeGenerator {
...
                 } else {
                     &folder_path_sdk
                 };
                 let filepath = folder_path.join(format!("{filename_stem}.{extension}"));
-                write_file(&filepath, &string);
-                let inserted = filepaths.insert(filepath);
+                let inserted = files_to_write.insert(filepath, contents);
                 assert!(
-                    inserted,
+                    inserted.is_none(),
```
Comment (emilk) [inline crates/re_types_builder/src/codegen/cpp/mod.rs:159]: `inserted` here is the wrong name - it should be `previous` or similar

Comment (emilk) [reply crates/re_types_builder/src/codegen/cpp/mod.rs:159]: (and GitHub wouldn't let me make a code suggestion for whatever reason)

File: crates/re_types_builder/src/lib.rs [hunk 2]
```diff
@@ -365,9 +379,41 @@ pub fn generate_cpp_code(
...
+    // 3. Write all files.
+    {
+        use rayon::prelude::*;
+
+        re_tracing::profile_scope!("write_files");
+
+        files.par_iter().for_each(|(filepath, contents)| {
+            // There's more than cpp/hpp files in here, don't run clang-format on them!
+            let contents =
+                if filepath.extension() == Some("cpp") || filepath.extension() == Some("hpp") {
```
Comment (emilk) [inline crates/re_types_builder/src/lib.rs:397]: style choice, but this is nice and short:

```suggestion
                if matches!(filepath.extension(), Some("cpp" | "hpp")) {
```

File: crates/re_types_builder/src/lib.rs [hunk 3]
```diff
@@ -392,14 +438,61 @@ pub fn generate_cpp_code(
...
+    crate::codegen::common::remove_orphaned_files(reporter, &files);
+
+    fn write_files(files: &GeneratedFiles) {
+        use rayon::prelude::*;
+
+        re_tracing::profile_function!();
+
+        // TODO(emilk): running `cargo fmt` once for each file is very slow.
+        // It would probably be faster to write all files to a temporary folder, run carg-fmt on
+        // that folder, and then copy the results to the final destination (if the files has changed).
```
Comment (emilk) [inline crates/re_types_builder/src/lib.rs:464]: Actually, I've tested this and it doesn't win us anything

```suggestion
```

File: crates/re_types_builder/src/lib.rs [hunk 4]
```diff
@@ -431,22 +524,133 @@ pub fn generate_python_code(
...
+        let output = proc.wait_with_output()?;
+
+        if output.status.success() {
+            Ok(())
+        } else {
+            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_owned();
+            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_owned();
+            anyhow::bail!("{stdout}\n{stderr}")
+        }
+    }
```
Comment (emilk) [inline crates/re_types_builder/src/lib.rs:653]: I don't understand why you moved out the language-specific formatting out of the language-specific modules?

To me it makes more sense if `GeneratedFiles` would be the final files, with no extra processing required.
Then these functions would be a lot more similar to each other (identical?)

Comment (emilk) [reply crates/re_types_builder/src/lib.rs:653]: …and it would make the PR diff smaller too :)

**Review by emilk — APPROVED** [review]

File: crates/re_types_builder/src/lib.rs (hunk 4 above)
Comment (teh-cmc) [reply crates/re_types_builder/src/lib.rs:649]: I just don't like the the mixup of concerns: the codegen pass should only concern itself with generating code, not any post-processing (including formatting) nor I/O.
Especially so with the current implementation where the codegen pass is nice and pure while the formatting pass requires I/O, process spawning, etc.

I will likely introduce a formalized formatting pass in https://github.com/rerun-io/rerun/issues/3495 though, which will make all these `generate_XXX` methods go away and make `--check`ing trivial.

Comment (emilk) [reply crates/re_types_builder/src/lib.rs:649]: that makes sense!

**Pull request merged**